prow/spyglass/spyglass_test.go

```diff
@@ -0,0 +1,186 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package spyglass
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	"cloud.google.com/go/storage"
+	"github.com/fsouza/fake-gcs-server/fakestorage"
+	"github.com/sirupsen/logrus"
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/deck/jobs"
+	"k8s.io/test-infra/prow/kube"
+)
+
+var (
+	fakeGCSBucket    *storage.BucketHandle
+	testAf           *GCSArtifactFetcher
+	fakeJa           *jobs.JobAgent
+	fakeGCSJobSource jobSource
+)
+
+const (
+	testSrc = "gs://test-bucket/logs/example-ci-run/403"
+)
+
+type fkc []kube.ProwJob
+
+func (f fkc) GetLog(pod string) ([]byte, error) {
+	return nil, nil
+}
+
+func (f fkc) ListPods(selector string) ([]kube.Pod, error) {
+	return nil, nil
+}
+
+func (f fkc) ListProwJobs(s string) ([]kube.ProwJob, error) {
+	return f, nil
+}
+
+type fpkc string
+
+func (f fpkc) GetLog(pod string) ([]byte, error) {
+	if pod == "wowowow" || pod == "powowow" {
+		return []byte(f), nil
+	}
+	return nil, fmt.Errorf("pod not found: %s", pod)
+}
+
+func (f fpkc) GetContainerLog(pod, container string) ([]byte, error) {
+	if pod == "wowowow" || pod == "powowow" {
+		return []byte(f), nil
+	}
+	return nil, fmt.Errorf("pod not found: %s", pod)
+}
+
+func (f fpkc) GetLogTail(pod, container string, n int64) ([]byte, error) {
+	if pod == "wowowow" || pod == "powowow" {
+		tailBytes := []byte(f)
+		lenTailBytes := int64(len(tailBytes))
+		if lenTailBytes < n {
+			return tailBytes, nil
+		}
+		return tailBytes[lenTailBytes-n-1:], nil
+	}
+	return nil, fmt.Errorf("pod not found: %s", pod)
+}
+
+func TestMain(m *testing.M) {
+	var longLog string
+	for i := 0; i < 300; i++ {
+		longLog += "here a log\nthere a log\neverywhere a log log\n"
+	}
+	fakeGCSServer := fakestorage.NewServer([]fakestorage.Object{
```

Comment:
Each test uses at most one of these little bits. Make them data-driven from the test. Let the test call a constructor and pass in string literals or whatever, where necessary. Drive the functionality of your fake from data in your test matrix. Don't see "duplicated" test fakes as an opportunity to DRY things out, as that starts creating weird dependencies between tests and starts to broaden the scope of your tests from "X functionality works" to "X functionality works and the shared test fake is set-up in the correct way". I should not be able to break test Y unless I break functionality Y (generic fake harnesses aside). See how the fake clients for k8s let you seed them with behavior at creation time.